### ISSUE TITLE
NPE occurs when the same type of ArgumentMatcher is used more than once

### ISSUE THREAD

**Issue by marchpig** [mockito/mockito#1442]
- [x] The mockito message in the stacktrace have useful information, but it didn't help
 - [x] The problematic code (if that's possible) is copied here;
       Note that some configuration are impossible to mock via Mockito
 - [x] Provide versions (mockito / jdk / os / any other relevant information)
 - [x] Provide a [Short, Self Contained, Correct (Compilable), Example](http://sscce.org) of the issue
       (same as any question on stackoverflow.com)
 - [x] Read the [contributing guide](https://github.com/mockito/mockito/blob/master/.github/CONTRIBUTING.md)

Hi,

I report an NPE issue.
The case is when the same type of `ArgumentMatcher` is used more than once in a test.
It happens with `when`.`thenReturn`, but not with `doReturn`.`when`.
Is it a know issue?
I'm recently reading the Mockito docs carefully but couldn't find anything about this case.
Please check the following test case and log.

```Java
private final class ContainsFoo implements ArgumentMatcher<String> { // Line 25
    public boolean matches(String arg) {
        return arg.contains("foo"); // Line 27
    }
}

private final class ContainsBar implements ArgumentMatcher<String> {
    public boolean matches(String arg) {
        return arg.contains("bar");
    }
}

@Test
public void foobarTest() {
    // NPE occurs
    when(mock.oneArg(argThat(new ContainsFoo()))).thenReturn("foo");
    when(mock.oneArg(argThat(new ContainsBar()))).thenReturn("bar"); // Line 41

    // Works fine
    // doReturn("foo").when(mock).oneArg(argThat(new ContainsFoo()));
    // doReturn("bar").when(mock).oneArg(argThat(new ContainsBar()));

    assertEquals("foo", mock.oneArg("foo"));
    assertEquals("bar", mock.oneArg("bar"));
}
```
```Java
java.lang.NullPointerException
	at org.mockitousage.matchers.CustomMatchersTest$ContainsFoo.matches(CustomMatchersTest.java:27)
	at org.mockitousage.matchers.CustomMatchersTest$ContainsFoo.matches(CustomMatchersTest.java:25)
	at org.mockito.internal.invocation.TypeSafeMatching.apply(TypeSafeMatching.java:24)
	at org.mockito.internal.invocation.MatcherApplicationStrategy.forEachMatcherAndArgument(MatcherApplicationStrategy.java:83)
	at org.mockito.internal.invocation.InvocationMatcher.argumentsMatch(InvocationMatcher.java:152)
	at org.mockito.internal.invocation.InvocationMatcher.matches(InvocationMatcher.java:81)
	at org.mockito.internal.stubbing.InvocationContainerImpl.findAnswerFor(InvocationContainerImpl.java:85)
	at org.mockito.internal.handler.MockHandlerImpl.handle(MockHandlerImpl.java:90)
	at org.mockito.internal.handler.NullResultGuardian.handle(NullResultGuardian.java:29)
	at org.mockito.internal.handler.InvocationNotifierHandler.handle(InvocationNotifierHandler.java:35)
	at org.mockito.internal.creation.bytebuddy.MockMethodInterceptor.doIntercept(MockMethodInterceptor.java:63)
	at org.mockito.internal.creation.bytebuddy.MockMethodInterceptor.doIntercept(MockMethodInterceptor.java:49)
	at org.mockito.internal.creation.bytebuddy.MockMethodInterceptor$DispatcherDefaultingToRealMethod.interceptAbstract(MockMethodInterceptor.java:128)
	at org.mockitousage.IMethods$MockitoMock$1726084737.oneArg(Unknown Source)
	at org.mockitousage.matchers.CustomMatchersTest.foobarTest(CustomMatchersTest.java:41)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

Thanks,
Sangwoo

**Comment by mockitoguy** [comment]
Your matcher implementation needs to handle null because matcher cannot assume that the "arg" will never be null.

You're getting an NPE because "when().then()" syntax actually invokes the stubbed method. It's a caveat of the API. In your second stubbing, you actually call method 'oneArg' that triggers the first stubbing to be evaluated. 1st stubbing has a matcher that does not handle nulls properly. All argument matching methods such as 'argThat' return null.

Hope that explanation makes sense! There's nothing we can fix about it.

**Comment by marchpig** [comment]
Thanks @mockitoguy for the explanation!
Now I understand a little bit how stubbing works.
I'm closing it.
Thanks!

**Comment by donce** [comment]
I've stumbled upon this issue as well and was quite confused, and didn't realize it's expected from the docs.

>You're getting an NPE because "when().then()" syntax actually invokes the stubbed method. It's a caveat of the API. In your second stubbing, you actually call method 'oneArg' that triggers the first stubbing to be evaluated.

Is this behavior documented somewhere? I've stumbled upon this issue and this issue was the most informative source of information.

I've also read "NullPointerException auto-unboxing caveat." in [ArgumentMatcher.argThat javadoc](https://javadoc.io/static/org.mockito/mockito-core/5.8.0/org/mockito/ArgumentMatchers.html#argThat(org.mockito.ArgumentMatcher)), but there's no primitives involved in my or this flow, so disregarded that.
